File: 10_DEALER/answer_2.py
import numpy as np


class DealerModel:
    def __init__(self, N, d, M, s, c, dp, dt, p0, max_tick=10000):
        self.N = N
        self.d = d
        self.M = M
        self.s = s
        self.c = c
        self.dp = dp
        self.dt = dt
        self.p0 = p0
        self.max_tick = max_tick
        self.mid_price_list = np.ones(N) * p0
        self.market_price_list = []

    @property
    def tick(self):
        return len(self.market_price_list)

    @property
    def best_ask(self):
        return np.min(self.mid_price_list) + self.s/2

    @property
    def best_bid(self):
        return np.max(self.mid_price_list) - self.s/2

    @property
    def best_ask_id(self):
        return np.argmin(self.mid_price_list)

    @property
    def best_bid_id(self):
        return np.argmax(self.mid_price_list)

    def transaction(self):
        deal_price = (self.best_ask + self.best_bid) / 2
        self.market_price_list.append(deal_price)
        self.mid_price_list[self.best_ask_id] = deal_price
        self.mid_price_list[self.best_bid_id] = deal_price
        print("\r""Tick:{}/{}, Price:{}".format(self.tick, self.max_tick, deal_price), end="")

    def random_walk(self):
        self.mid_price_list += self.c * np.random.choice([self.dp, -self.dp], self.N)

    def trend_follow(self):
        dp_list = np.diff(self.market_price_list)
        n = len(dp_list) if self.M >= len(dp_list) else self.M
        trend = sum([(2 * dp_list[-(k + 1)] * (self.M - k)) / (self.M * (self.M + 1)) for k in range(n)])
        self.mid_price_list += self.d * self.dt * trend

    def run(self):
        while True:
            if self.tick >= self.max_tick:
                break
            if self.best_ask < self.best_bid:
                self.transaction()
                continue
            self.random_walk()
            self.trend_follow()


if __name__ == "__main__":
    np.random.seed(1129)
    model = DealerModel(N=2, d=0, M=2, s=0.01, c=0.01, dp=0.1, dt=0.1*0.1, p0=100)
    model.run()
    market_price_list = model.market_price_list

File: 10_DEALER/cpp/answer_2.cpp
#include <stdio.h>
#include <vector>
#include <random>
#include <fstream>
#include <string>

std::mt19937 mt(1129);
std::uniform_real_distribution<> uni(0.0, 1.0);
std::string PRICE_CSV = "price_2.csv";

class DealerModel
{
public:
    std::vector<double> _market_price_vector;
    DealerModel(int N, double s, double c, double d, double dp, double dt, int M, double p0, int max_tick = 10000)
    {
        _N = N;
        _s = s;
        _c = c;
        _d = d;
        _dp = dp;
        _dt = dt;
        _dt = dt;
        _M = M;
        _p0 = p0;
        _tick = 0;
        _max_tick = max_tick;
        for (int i = 0; i < N; i++)
            _mid_price_vector.push_back(p0);
    }
    // method
    void run();

private:
    int _N;
    double _s;
    double _c;
    double _d;
    double _dp;
    double _dt;
    int _M;
    double _p0;
    int _max_tick;
    int _tick;
    std::vector<double> _mid_price_vector;
    // method
    int transaction();
    void trend_follow_random_walk();
};

void DealerModel::trend_follow_random_walk()
{
    int n;
    double trend = 0;
    if (_tick - 1 < _M)
        n = _tick - 1;
    else
        n = _M;
    for (int k = 0; k < n; k++)
    {
        double dP = _market_price_vector[(_tick - 1) - k] - _market_price_vector[(_tick - 1) - k - 1];
        trend += 2 * dP * (_M - k) / (_M * (_M + 1));
    }
    for (int i = 0; i < _N; i++)
    {
        if (uni(mt) < 0.5)
            _mid_price_vector[i] += _d * trend * _dt + _dp * _c;
        else
            _mid_price_vector[i] += _d * trend * _dt - _dp * _c;
    }
}

int DealerModel::transaction()
{
    double best_ask = _mid_price_vector[0] + _s / 2;
    double best_bid = _mid_price_vector[0] - _s / 2;
    int best_ask_idx = 0;
    int best_bid_idx = 0;
    for (int i = 0; i < _N; i++)
    {
        if (_mid_price_vector[i] + _s / 2 < best_ask)
        {
            best_ask = _mid_price_vector[i] + _s / 2;
            best_ask_idx = i;
        }
        else if (_mid_price_vector[i] - _s / 2 > best_bid)
        {
            best_bid = _mid_price_vector[i] - _s / 2;
            best_bid_idx = i;
        }
    }
    if (best_ask <= best_bid)
    {
        double market_price = (best_ask + best_bid) / 2;
        _market_price_vector.push_back(market_price);
        _mid_price_vector[best_ask_idx] = market_price;
        _mid_price_vector[best_bid_idx] = market_price;
        _tick++;
        printf("tick:%d, price:%f\r", _tick, market_price);
        return 1;
    }
    else
        return 0;
}

void DealerModel::run()
{
    while (1)
    {
        if (transaction())
            continue;
        else
            trend_follow_random_walk();
        if (_tick >= _max_tick)
        {
            printf("%s", "END");
            break;
        }
    }
}

int main()
{
    int N = 2;
    double s = .01;
    double c = .01;
    double d = 1.25;
    double dp = .01;
    double dt = .01 * .01;
    int M = 1;
    double p0 = 100.0;
    int max_tick = 10000;
    DealerModel model(N, s, c, d, dp, dt, M, p0);
    model.run();
    std::ofstream price_csv(PRICE_CSV);
    price_csv << "tick,market_price" << std::endl;
    for (int i = 0; i < max_tick; i++)
    {
        price_csv << i + 1 << "," << model._market_price_vector[i] << std::endl;
    }
}